src/Templating/Helper/LocaleHelper.php

```diff
@@ -33,13 +51,19 @@ class LocaleHelper extends BaseHelper
      */
     public function country($code, $locale = null)
     {
-        // NEXT_MAJOR: Remove this when dropping < 4.3 Symfony support
-        if (!class_exists(Countries::class)) {
-            $name = Intl::getRegionBundle()->getCountryName($code, $locale ?: $this->localeDetector->getLocale());
-
-            return $name ? $this->fixCharset($name) : '';
+        if ($this->intlExtension) {
+            return $this->fixCharset($this->intlExtension->getCountryName($code, $locale ?: $this->localeDetector->getLocale()));
         }
 
+        // NEXT_MAJOR: Execute the previous block unconditionally and remove following lines in this method.
+
+        @trigger_error(sprintf(
+            'Not passing an instance of "%s" as argument 3 for %s::__construct() is deprecated since sonata-project/intl-bundle 2.x.'
+            .' and will throw an exception since version 3.x.',
+            IntlExtension::class,
+            __CLASS__
+        ));
```

Comment:
Why not have this once in the constuctor?

Reply:
Because this class can be extended, and the inherited class could override `__construct()` without calling its parent.

Reply:
It's dangerous to override the constructor without calling the original one. We should risk that case, because it's his own fault.

Reply:
>It's dangerous to override the constructor without calling the original one.

I agree that premise. Our BC promise and our current API is causing a lot of situations which we have to "adivinate" in order to prevent a bad DX.
My point is, we havent an explicit path for these cases, so the same reasoning could apply for method arguments which has declared its allowed types with a docblock:

>It's dangerous to pass other types than allowed by the method's docblocks.

Nevertheless, we are currently spending our effort on that possible situtations.

Reply:
What does "adivinate" mean exactly? I don't know that word.

Reply:
I think it should be: guess.

Agree on moving the deprecation to the constructor, if you are extending this class, normally you would call the parent constructor and the deprecation is happening there, because we added new param. We have other cases like that already.